lib/SILOptimizer/IPO/GlobalOpt.cpp

```diff
@@ -966,27 +1090,104 @@ bool SILGlobalOpt::run() {
           continue;
         }
 
-        auto *GAI = dyn_cast<GlobalAddrInst>(&I);
-        if (!GAI) {
+        if (auto *GAI = dyn_cast<GlobalAddrInst>(&I)) {
+          collectGlobalAccess(GAI);
           continue;
         }
 
-        collectGlobalAccess(GAI);
+        if (auto *allocGlobal = dyn_cast<AllocGlobalInst>(&I)) {
+          collectAllocGlobal(allocGlobal->getReferencedGlobal(), allocGlobal);
+          continue;
+        }
       }
     }
   }
+}
+
+bool SILGlobalOpt::run() {
+  // Collect all the global variables and associated instructions.
+  collect();
 
+  // Optimize based on what we just collected.
   for (auto &InitCalls : GlobalInitCallMap) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    bool shouldOptimize = true;
+    for (auto *apply : InitCalls.second) {
+      if (!apply->getFunction()->shouldOptimize()) {
+        shouldOptimize = false;
+        break;
+      }
+    }
+    if (!shouldOptimize)
+      continue;
+
     // Optimize the addressors if possible.
     optimizeInitializer(InitCalls.first, InitCalls.second);
     placeInitializers(InitCalls.first, InitCalls.second);
   }
 
   for (auto &Init : GlobalVarStore) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    if (!Init.second->getFunction()->shouldOptimize())
+      continue;
+
     // Optimize the access to globals if possible.
     optimizeGlobalAccess(Init.first, Init.second);
   }
 
+  SmallVector<SILGlobalVariable *, 8> addrGlobals;
+  for (auto &addrPair : GlobalAddrMap) {
+    // Don't optimize functions that are marked with the opt.never attribute.
+    bool shouldOptimize = true;
+    for (auto *addr : addrPair.second) {
+      if (!addr->getFunction()->shouldOptimize()) {
+        shouldOptimize = false;
+        break;
+      }
+    }
+    if (!shouldOptimize)
+      continue;
+
+    addrGlobals.push_back(addrPair.first);
+  }
+
+  for (auto *global : addrGlobals) {
+    HasChanged |= tryRemoveGlobalAddr(global);
+  }
+
+  SmallVector<std::pair<SILGlobalVariable *, AllocGlobalInst *>, 12>
+      globalAllocPairs;
+  for (auto &alloc : AllocGlobalStore) {
+    if (!alloc.second->getFunction()->shouldOptimize())
+      continue;
+    globalAllocPairs.push_back(std::make_pair(alloc.first, alloc.second));
+  }
+
+  for (auto &allocPair : globalAllocPairs) {
+    HasChanged |= tryRemoveGlobalAlloc(allocPair.first, allocPair.second);
+  }
+
+  // Erase the instructions that we have marked for deletion.
+  for (auto *inst : InstToRemove) {
+    inst->eraseFromParent();
+  }
+
+  // After we erase some instructions, re-collect.
+  reset();
+  collect();
```

Comment:
If the second call to collect() is just in the service of tryRemoveUnusedGlobal(), it would be wiser to split up the analysis into two parts, with the second part only running after instructions have been erased, as above.

Reply:
`tryRemoveUnusedGlobal()` checks `GlobalAddrMap`, `GlobalAccessMap`, `GlobalLoadMap`, etc. so these need to be re-collected. Otherwise, it wouldn't be able to optimize away a global variable in one pass.

If SILGlobalVariable had a `getUses` method, we wouldn't need to recollect. I'm not sure if there is a reason it doesn't, but, I'll look into making a patch to add one.

Reply:
Alternatively, I could check that the global isn't in those maps (`GlobalAddrMap`, `GlobalAccessMap`, `GlobalLoadMap`, etc.) or that all values (corresponding to the global) are an instruction in `InstToRemove`. It would be a messier but probably faster.

Reply:
The last approach is what I had in mind. You might find `std::all_of()` helpful to collapse the innermost loop.